dvc/remote/base.py

```diff
@@ -47,123 +90,393 @@ def __repr__(self):
             path_info=self.tree.path_info or "No path",
         )
 
-    @property
-    def cache(self):
-        return getattr(self.repo.cache, self.tree.scheme)
-
-    def hashes_exist(self, hashes, jobs=None, name=None):
-        """Check if the given hashes are stored in the remote.
-
-        There are two ways of performing this check:
-
-        - Traverse method: Get a list of all the files in the remote
-            (traversing the cache directory) and compare it with
-            the given hashes. Cache entries will be retrieved in parallel
-            threads according to prefix (i.e. entries starting with, "00...",
-            "01...", and so on) and a progress bar will be displayed.
-
-        - Exists method: For each given hash, run the `exists`
-            method and filter the hashes that aren't on the remote.
-            This is done in parallel threads.
-            It also shows a progress bar when performing the check.
-
-        The reason for such an odd logic is that most of the remotes
-        take much shorter time to just retrieve everything they have under
-        a certain prefix (e.g. s3, gs, ssh, hdfs). Other remotes that can
-        check if particular file exists much quicker, use their own
-        implementation of hashes_exist (see ssh, local).
-
-        Which method to use will be automatically determined after estimating
-        the size of the remote cache, and comparing the estimated size with
-        len(hashes). To estimate the size of the remote cache, we fetch
-        a small subset of cache entries (i.e. entries starting with "00...").
-        Based on the number of entries in that subset, the size of the full
-        cache can be estimated, since the cache is evenly distributed according
-        to hash.
-
-        Returns:
-            A list with hashes that were found in the remote
-        """
-        # Remotes which do not use traverse prefix should override
-        # hashes_exist() (see ssh, local)
-        assert self.tree.TRAVERSE_PREFIX_LEN >= 2
+    def gc(self, *args, **kwargs):
+        removed = self.cache.gc(*args, **kwargs)
+
+        if removed:
+            with self.index:
+                self.index.clear()
+
+        return removed
+
+    @index_locked
+    def status(
+        self,
+        cache,
+        named_cache,
+        jobs=None,
+        show_checksums=False,
+        download=False,
+        log_missing=True,
+    ):
+        # Return flattened dict containing all status info
+        dir_status, file_status, _ = self._status(
+            cache,
+            named_cache,
+            jobs=jobs,
+            show_checksums=show_checksums,
+            download=download,
+            log_missing=log_missing,
+        )
+        return dict(dir_status, **file_status)
 
+    def hashes_exist(self, hashes: set, **kwargs):
         hashes = set(hashes)
```

Comment:
This conversion that we do might be causing things like https://github.com/iterative/dvc/issues/4139 . Might want to check.